### PR TITLE
Half Demo Discount Api

### PR DESCRIPTION
**Pull request by LennyDuan**
写了一些测试，api还没有实现。。。

纠结了一晚上。。发现promise还是有点蒙。。。

你看那个被我TODO comment 的那个promise，还是有点蒙 :(
exports.isBelongToUsers
我得研究一下Promise这个东西....

### CODE REVIEW

File: proxy/coupon.model.js [hunk 1]
```diff
@@ -33,4 +34,34 @@ exports.createCouponWithDefaultRulesForSpecifiedUser = function (username, coupo
...
+};
+
+// TODO how to write Promise + Promise
+// exports.isBelongToUsers = function (couponId, username) {
+//   var couponPromise = getCouponCodesByCouponCode(couponId);
+//   var usersCouponsPromise = getCouponCodesByUsername(username);
+//   return Promise.join(couponPromise, usersCouponsPromise, function(coupon, couponArray)){
+//     return Arrays.asList(couponArray).contains(coupon);
+//   }
+// }
```
Comment (realfangren) [inline proxy/coupon.model.js]: 直接一个promise就好了. Coupon model里面有username嘛

return getCouponCodesByCouponCode(couponId).then(function(coupons){
  //返回的是数组，但是理论上只会有一个coupon所以直接取第一个
   return coupons[0].username == username;
})

**Comment by LennyDuan** [comment]
bingo

File: api/v1/coupons.js [hunk 2]
```diff
@@ -30,6 +32,38 @@ var createCouponForNewUser = function (req, res, next) {
...
+        console.log(ruleValue + '---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------')
+        console.log(orderValue + '---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------')
+
+        var discountedValue = coupon.getDiscountedValue(ruleType, ruleValue, orderValue);
+        res.statusCode = 201;
```
Comment (LennyDuan) [inline api/v1/coupons.js:62]: 这里有个bug修了半天....
Error: Can't set headers after they are sent.
应该是异步的问题？
明天再看.....

Comment (LennyDuan) [reply api/v1/coupons.js:62]: console 是为了debug......以下能找到

Comment (realfangren) [reply api/v1/coupons.js:62]: yes, you'd need to chain your promises. For example

```
Promise.join(A, B, functon(retA, retB){
.....
return true
}).then(function(ableToUse){
  ...
 return someValue
}).then(function(someValue){
  res.send(someValue)
}).catch(...)

```

Also you have not handle the case where coupon is not valid.

Comment (realfangren) [reply api/v1/coupons.js:62]: chained promises are not nested, they would look like this:

```
remotedb.allDocs(...).then(function (resultOfAllDocs) {
  return localdb.put(...);
}).then(function (resultOfPut) {
  return localdb.get(...);
}).then(function (resultOfGet) {
  return localdb.put(...);
}).catch(function (err) {
  console.log(err);
});

```

And this is the same code but a bad example

```
remotedb.allDocs({
  include_docs: true,
  attachments: true
}).then(function (result) {
  var docs = result.rows;
  docs.forEach(function(element) {
    localdb.put(element.doc).then(function(response) {
      alert("Pulled doc with id " + element.doc._id + " and added to local db.");
    }).catch(function (err) {
      if (err.name == 'conflict') {
        localdb.get(element.doc._id).then(function (resp) {
          localdb.remove(resp._id, resp._rev).then(function (resp) {
```

Have a read at this https://pouchdb.com/2015/05/18/we-have-a-problem-with-promises.html

Comment (realfangren) [reply api/v1/coupons.js:62]: However, this is a style problem, the real source for your error (Error: Can't set headers after they are sent.) is the
- `res.send('TODO');`line 64, is sent outside the promise, hence you have send the res before the promises returned, and then you set the header in the promise return     `res.statusCode = 201;.` at line 58

File: middlewares/discountChecker.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  }
+
+  if(!orderValue) {
+  	var err = new Error('No orderValue Provided');
+    err.status = 406;
+    next(err);
+  }
+  next();
+};
+
```
Comment (realfangren) [inline middlewares/discountChecker.js:18]: this middleware validates username and order value however, it's unlikely that these checks will be used by other routers, why not putting the checks into the router?

File: proxy/coupon.model.js [hunk 4]
```diff
@@ -33,4 +34,45 @@ exports.createCouponWithDefaultRulesForSpecifiedUser = function (username, coupo
...
     couponRule : defaultRules.couponRule, rebateRule : defaultRules.rebateRule};
   return new Coupon(newCoupon).save();
+};
+
+exports.isBelongToUsers = function (couponId, username) {
+  return Coupon.find({ couponID : couponId} ).then (function (coupons){
+    return coupons[0].username === username;
+  }, function (err) {
+    return err;
+  });
```
Comment (realfangren) [inline proxy/coupon.model.js:48]: don't use this style use catch or let the caller catches it.

File: proxy/coupon.model.js [hunk 5]
```diff
@@ -33,4 +34,45 @@ exports.createCouponWithDefaultRulesForSpecifiedUser = function (username, coupo
...
+    dicountedValue = parseInt(dicountedValue);
+    console.log(dicountedValue + '---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------')
+
+    return  { 'couponID': couponObject.couponID, 'dicountedValue' : dicountedValue };
+
+  }).then(function (discountedOrder) {
+    console.log(discountedOrder + '---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------')
+    return discountedOrder;
+  });
 };
```
Comment (realfangren) [inline proxy/coupon.model.js:86]: I'd think this is part of the business logic, maybe it's better move to order to the router, but it's up to you.

**Comment by LennyDuan** [comment]
我特么的醉了了。。。我测试写错了。。。我怎么说调了半天怎么调都是错的。。。
其实我昨天晚上就应该写对了╮(╯▽╰)╭ 忧伤的故事

**Pull request merged**